cezo_fl/client.py

```diff
@@ -0,0 +1,138 @@
+import torch
+from typing import Sequence
+from copy import deepcopy
+
+from gradient_estimators.random_gradient_estimator import RandomGradientEstimator as RGE
+from torch.optim import SGD
+from cezo_fl.server import AbstractClient
+
+
+class Client(AbstractClient):
+
+    def __init__(
+        self,
+        model,
+        dataloader,
+        device,
+        grad_estimator_params: dict,
+        optimzier_params: dict,
+        criterion,
+    ):
+        self.model = model
+        self.dataloader = dataloader
+
+        self.device = device
+
+        self.grad_estimator = self._initialize_grad_estimator(grad_estimator_params)
+
+        self.optimizer = self._initialize_optimizer(optimzier_params)
+        self.criterion = criterion
+
+        self.data_iterator = self._get_train_batch_iterator()
+        self.last_pull_state_dict = self.screenshot()
+
+    def _initialize_grad_estimator(self, grad_estimator_params):
+        if grad_estimator_params["method"] in ["rge-forward", "rge-central"]:
+            method = grad_estimator_params["method"][4:]
+
+            return RGE(
+                self.model,
+                grad_estimate_method=method,
+                mu=grad_estimator_params["mu"],
+                num_pert=grad_estimator_params["num_pert"],
+                device=self.device,
+            )
+        else:
+            raise Exception(f'{grad_estimator_params["method"]} is not supported')
+
+    def _initialize_optimizer(self, optimzier_params):
```

Comment:
I prefer this initialization put at the main file and just pass the initialized optimizer into this class. Notice we need to make sure the optimizer in clients and server are the same